### ISSUE TITLE
Split hyper from iron core

### ISSUE THREAD

**Issue by sunng87** [iron/iron#261]
Currently, our http server hyper is built into core parts of Iron. If you look at Python's WSGI or Clojure's Ring, there is a protocol between webapp and web server. And the server implementation are made separated from the protocol and webapp. In Clojure, for example, the protocol is [ring](https://github.com/ring-clojure/ring), it's middleware based like Iron. It comes with a default server implementation ring-jetty-adapter, but it's totally decoupled from ring, so you can write servers and make adapters for ring. In python, there's also similar framework like bottle.

Although hyper is the only http server currently, I think it's still necessary to split and make each part easy to test.

**Comment by reem** [comment]
I agree completely. I haven't really felt the need to do this yet since everything is still in flux and hyper is the only real option right now, but eventually I think there should be a protocol/trait layer between hyper and Iron so that you can swap out the hyper back end.
